File: api/src/main/java/net/jqwik/api/arbitraries/SizableArbitrary.java
package net.jqwik.api.arbitraries;

import org.apiguardian.api.*;

import net.jqwik.api.*;

import org.jspecify.annotations.*;

import static org.apiguardian.api.API.Status.*;

/**
 * Fluent interface to configure arbitraries that have size constraints for generated values, e.g. collections and arrays.
 */
@API(status = MAINTAINED, since = "1.0")
public interface SizableArbitrary<U extends @Nullable Object> extends Arbitrary<U> {

	/**
	 * Fix the size to {@code size}.
	 */
	default SizableArbitrary<U> ofSize(int size) {
		return ofMinSize(size).ofMaxSize(size);
	}

	/**
	 * Set lower size boundary {@code minSize} (included).
	 */
	SizableArbitrary<U> ofMinSize(int minSize);

	/**
	 * Set upper size boundary {@code maxSize} (included).
	 */
	SizableArbitrary<U> ofMaxSize(int maxSize);

	/**
	 * Set distribution {@code distribution} of size of generated arbitrary
	 */
	@API(status = EXPERIMENTAL, since = "1.5.3")
	SizableArbitrary<U> withSizeDistribution(RandomDistribution distribution);
}

File: api/src/main/kotlin/net/jqwik/api/arbitraries/SizableArbitrary.kt
package net.jqwik.api.arbitraries

import net.jqwik.api.Arbitrary
import net.jqwik.api.RandomDistribution
import org.apiguardian.api.API

/**
 * Fluent interface to configure arbitraries that have size constraints for generated values, e.g. collections and arrays.
 */
@API(status = API.Status.MAINTAINED, since = "1.0")
interface SizableArbitrary<out U> : Arbitrary<U> {
    /**
     * Fix the size to `size`.
     */
    fun ofSize(size: Int): SizableArbitrary<U>? {
        return ofMinSize(size).ofMaxSize(size)
    }

    /**
     * Set lower size boundary `minSize` (included).
     */
    fun ofMinSize(minSize: Int): SizableArbitrary<U>

    /**
     * Set upper size boundary `maxSize` (included).
     */
    fun ofMaxSize(maxSize: Int): SizableArbitrary<U>?

    /**
     * Set distribution `distribution` of size of generated arbitrary
     */
    @API(status = API.Status.EXPERIMENTAL, since = "1.5.3")
    fun withSizeDistribution(distribution: RandomDistribution): SizableArbitrary<U>
}